File: assets/rt/flows/pipelineParallel/ConvertFromDataParallel.java
package rt.flows.pipelineParallel;

import base.*;
import base.flows.FlowOp_1;
import base.flows.Flow_1;
import base.flows._PipelineParallelFlow_0;
import base.flows._Sink_1;
import rt.flows.dataParallel.DataParallelFlow;

import java.util.concurrent.BlockingDeque;
import java.util.concurrent.LinkedBlockingDeque;

/**
 * Certain operations are not safe in DP, so we need to convert to PP.
 * However, if we just start adapting the flow-op from that point,
 * everything before the operator that made us convert to DP will be
 * sequential-- which is sad.
 * This operation will effectively split the flow up into two flows:
 * A DP flow (1) with a terminal operator that runs a PP flow (2).
 */
public interface ConvertFromDataParallel {
  static Flow_1 of(DataParallelFlow source, Opt_1 size) {
    var dpSource = source.getDataParallelSource();
    var dpConsumer = new DPSource(dpSource);

    return _PipelineParallelFlow_0.$self.fromOp$imm(dpConsumer, size);
  }
}

class DPSource implements FlowOp_1 {
  private final BlockingDeque<Object> buffer = new LinkedBlockingDeque<>();
  private boolean isRunning = true;
  private boolean hasStarted = false;
  private final DataParallelFlow.DataParallelSource source;
  DPSource(DataParallelFlow.DataParallelSource source) {
    this.source = source;
  }

  private void start() {
    assert !hasStarted : "start called twice on DPSource";
    hasStarted = true;
    Thread.ofVirtual().start(()->
      source.for$mut(new _Sink_1() {
        @Override public Void_0 stopDown$mut() {
          buffer.offer(PipelineParallelFlow.Message.Stop.INSTANCE);
          return Void_0.$self;
        }
        @Override public Void_0 pushError$mut(Info_0 info_m$) {
          buffer.offer(new PipelineParallelFlow.Message.Error(info_m$));
          return Void_0.$self;
        }
        @Override public Void_0 $hash$mut(Object x_m$) {
          buffer.offer(x_m$);
          return Void_0.$self;
        }
      }));
  }

  @Override public Bool_0 isFinite$mut() {
    return source.isFinite$mut();
  }
  @Override public Void_0 step$mut(_Sink_1 sink_m$) {
    if (isRunning$mut() == False_0.$self) {
      return Void_0.$self;
    }
    if (!hasStarted) {
      start();
    }
    var sink = (PipelineParallelFlow.WrappedSink) sink_m$;
    try {
      var msg = buffer.take();
      sink.subject.submit(msg);
      if (msg == PipelineParallelFlow.Message.Stop.INSTANCE) {
        isRunning = false;
      }
    } catch (InterruptedException e) {
      throw new RuntimeException(e);
    }
    if (!isRunning) {
      sink.stopDown$mut();
      this.stopUp$mut();
    }
    return Void_0.$self;
  }
  @Override public Void_0 stopUp$mut() {
    isRunning = false;
    source.stopUp$mut();
    return Void_0.$self;
  }
  @Override public Bool_0 isRunning$mut() {
    return isRunning ? True_0.$self : False_0.$self;
  }
  @Override public Void_0 for$mut(_Sink_1 downstream_m$) {
    return base.flows.FlowOp_1.for$mut$fun(downstream_m$, this);
  }
  @Override public Opt_1 split$mut() {
    return Opt_1.$self;
  }
  @Override public Bool_0 canSplit$read() {
    return False_0.$self;
  }
}

File: assets/rt-js/flows/pipelineParallel/ConvertFromDataParallel.js
// rt/flows/pipelineParallel/ConvertFromDataParallel.js
import { base$$flows$$_PipelineParallelFlow_0 } from "../../../base/flows/_PipelineParallelFlow_0.js";
import { base$$flows$$FlowOp_1 } from "../../../base/flows/FlowOp_1.js";
import { base$$True_0 } from "../../../base/True_0.js";
import { base$$False_0 } from "../../../base/False_0.js";
import { base$$Opt_1 } from "../../../base/Opt_1.js";
import { base$$Void_0 } from "../../../base/Void_0.js";
import { PipelineParallelFlow } from "./PipelineParallelFlow.js";

/* ---------------- ConvertFromDataParallel ---------------- */
export const ConvertFromDataParallel = {
  /**
   * @param {DataParallelFlow} source
   * @param {base$$Opt_1} size
   */
  of(source, size) {
    const dpSource = source.getDataParallelSource();
    const dpConsumer = new DPSource(dpSource);
    return base$$flows$$_PipelineParallelFlow_0.$self.fromOp$imm$2(
      dpConsumer,
      size
    );
  }
};

/* ---------------- DPSource ---------------- */
class DPSource {
  constructor(source) {
    this.buffer = [];
    this.waiters = [];
    this.isRunning = true;
    this.hasStarted = false;
    this.source = source;

    this._deliveryScheduled = false;

    // NEW: once true, no more data/errors are forwarded after Stop
    this._stopped = false;
  }

  async start() {
    if (this.hasStarted) throw new Error("start called twice on DPSource");
    this.hasStarted = true;

    (async () => {
      await this.source.for$mut$1({
        stopDown$mut$0: () => {
          // Mark hard-stopped, then enqueue Stop exactly once
          if (!this._stopped) {
            this._stopped = true;
            this._offer(PipelineParallelFlow.Message.Stop.INSTANCE);
          }
          return base$$Void_0.$self;
        },
        pushError$mut$1: info => {
          // If you want: stop the bridge on error too
          if (this._stopped) return base$$Void_0.$self;
          this._offer(new PipelineParallelFlow.Message.Error(info));
          return base$$Void_0.$self;
        },
        $hash$mut$1: x => {
          // Drop data after Stop
          if (this._stopped) return base$$Void_0.$self;
          this._offer(x);
          return base$$Void_0.$self;
        }
      });
    })();
  }

  _offer(item) {
    if (this.waiters.length > 0) this.waiters.shift()(item);
    else this.buffer.push(item);
  }
  async _take() {
    if (this.buffer.length > 0) return this.buffer.shift();
    return new Promise(r => this.waiters.push(r));
  }
  _poll() { return this.buffer.length > 0 ? this.buffer.shift() : undefined; }

  isFinite$mut$0() { return this.source.isFinite$mut$0(); }

  // Synchronous, non-blocking step (as you have now)
  step$mut$1(sink_m$) {
    if (this.isRunning$mut$0() === base$$False_0.$self) return base$$Void_0.$self;
    if (!this.hasStarted) this.start();

    const msg = this._poll();
    if (msg !== undefined) {
      sink_m$.subject.submit(msg);
      if (msg === PipelineParallelFlow.Message.Stop.INSTANCE) {
        this.isRunning = false;
        this.stopUp$mut$0();
      }
      return base$$Void_0.$self;
    }

    if (!this._deliveryScheduled) {
      this._deliveryScheduled = true;
      (async () => {
        try {
          const m = await this._take();
          sink_m$.subject.submit(m);
          if (m === PipelineParallelFlow.Message.Stop.INSTANCE) {
            this.isRunning = false;
            this.stopUp$mut$0();
          }
        } finally {
          this._deliveryScheduled = false;
        }
      })().catch(() => {});
    }

    return base$$Void_0.$self;
  }

  stopUp$mut$0() {
    this.isRunning = false;
    this.source.stopUp$mut$0();
    return base$$Void_0.$self;
  }
  isRunning$mut$0() { return this.isRunning ? base$$True_0.$self : base$$False_0.$self; }
  for$mut$1(downstream_m$) { return base$$flows$$FlowOp_1.for$mut$2$fun(downstream_m$, this); }
  split$mut$0() { return base$$Opt_1.$self; }
  canSplit$read$0() { return base$$False_0.$self; }
}
